protobuf_serialization.nim

```diff
@@ -0,0 +1,376 @@
+import macros, strformat, typetraits, options
+import faststreams
+
+template sint32*() {.pragma.}
+template sint64*() {.pragma.}
+template sfixed32*() {.pragma.}
+template sfixed64*() {.pragma.}
+template fixed32*() {.pragma.}
+template fixed64*() {.pragma.}
+template float*() {.pragma.}
+template double*() {.pragma.}
+
+const
+  MaxMessageSize* = 1'u shl 22
+
+type
+  ProtoBuffer* = object
+    fieldNum: int
+    outstream: OutputStreamVar
+
+  ProtoWireType* = enum
+    ## Protobuf's field types enum
+    Varint, Fixed64, LengthDelimited, StartGroup, EndGroup, Fixed32
+
+  EncodingKind* = enum
+    ekNormal, ekZigzag
+
+  ProtoField*[T] = object
+    ## Protobuf's message field representation object
+    index*: int
+    value*: T
+
+  SomeSVarint* = int | int64 | int32 | int16 | int8 | enum
+  SomeByte* = byte | bool | char | uint8
+  SomeUVarint* = uint | uint64 | uint32 | uint16 | SomeByte
+  SomeVarint* = SomeSVarint | SomeUVarint
+  SomeLengthDelimited* = string | seq[SomeByte] | cstring
+  SomeFixed64* = float64
+  SomeFixed32* = float32
+  SomeFixed* = SomeFixed32 | SomeFixed64
+
+  AnyProtoType* = SomeVarint | SomeLengthDelimited | SomeFixed | object
+
+  UnexpectedTypeError* = object of ValueError
+
+proc newProtoBuffer*(): ProtoBuffer =
+  ProtoBuffer(outstream: OutputStream.init(), fieldNum: 1)
+
+proc output*(proto: ProtoBuffer): seq[byte] {.inline.} =
+  proto.outstream.getOutput
+
+template wireType(firstByte: byte): ProtoWireType =
+  (firstByte and 0b111).ProtoWireType
+
+template fieldNumber(firstByte: byte): int =
+  ((firstByte shr 3) and 0b1111).int
+
+template protoHeader*(fieldNum: int, wire: ProtoWireType): byte =
+  ## Get protobuf's field header integer for ``index`` and ``wire``.
+  ((cast[uint](fieldNum) shl 3) or cast[uint](wire)).byte
+
+template increaseBytesRead(amount = 1) =
+  ## Convenience template for increasing
+  ## all of the counts
+  mixin isSome
+  bytesRead += amount
+  outOffset += amount
+  outBytesProcessed += amount
+  if numBytesToRead.isSome():
+    if (bytesRead > numBytesToRead.get()).unlikely:
+      raise newException(Exception, &"Number of bytes read ({bytesRead}) exceeded bytes requested ({numBytesToRead})")
+
+proc encodeField*[T: not AnyProtoType](protobuf: var ProtoBuffer, value: T) {.inline.}
+proc encodeField*[T: not AnyProtoType](protobuf: var ProtoBuffer, fieldNum: int, value: T) {.inline.}
+proc encodeField[T: not AnyProtoType](stream: OutputStreamVar, fieldNum: int, value: T) {.inline.}
+
+proc put(stream: OutputStreamVar, value: SomeVarint) {.inline.} =
+  when value is enum:
+    var value = cast[type(ord(value))](value)
+  elif value is bool or value is char:
+    var value = cast[byte](value)
+  else:
+    var value = value
+
+  when type(value) is SomeSVarint:
```

Comment:
this is not a give - ie there's no "default" in protobuf that "prefers" zig-zag - there are two types in the language, `intXX` and `sintXX` to differentiate - the framework here should not really have an opinion on which one is used (depends on the situation), but above all, it shouldn't surprise the user and choose one at random (ie I write `int32` in nim and get `sint32` semantics in protobuf). for signed integers, I'd say the cleanest thing to do would be to require an annotation in the type to choose between the two (unsigned integers are unambiguous so they don't need this)

Reply:
Yes, I already discussed this with @zah. As you can see, at the top of the file I have some annotations but was unsure how to use them. He pointed me in the right direction and I will be implementing it soon.